Bigtable/src/Instance.php

```diff
@@ -0,0 +1,362 @@
+            throw new \InvalidArgumentException('At least one clusterMetadata must be passed');
+        }
+        $projectName = InstanceAdminClient::projectName($this->projectId);
+        $displayName = $this->pluck('displayName', $options, false) ?: $this->id;
+        $labels = $this->pluck('labels', $options, false) ?: [];
+        $type = $this->pluck('type', $options, false);
+
+        $clustersArray = [];
+        foreach ($clusterMetadataList as $value) {
+            if (!isset($value['clusterId'])) {
+                throw new \InvalidArgumentException('Cluster id must be set.');
+            }
+            $this->validate($value['clusterId'], 'cluster');
+            $clusterId = $value['clusterId'];
+            if (!isset($value['locationId'])) {
+                throw new \InvalidArgumentException('Location id must be set.');
+            }
+            $this->validate($value['locationId'], 'location');
+            $locationId = $value['locationId'];
+            $value['location'] = InstanceAdminClient::locationName($this->projectId, $locationId);
+            if ($type === Instance_Type::DEVELOPMENT) {
+                unset($value['serveNodes']);
+            } elseif ($type === Instance_Type::PRODUCTION
+                        && (!isset($value['serveNodes']) || $value['serveNodes'] <= 0)) {
+                throw new \InvalidArgumentException('When creating Production instance, serveNodes must be > 0');
+            }
+            // `$clustersArray` must be keyed by the cluster ID.
+            $clustersArray[$clusterId] = $value;
+        }
+        $operation = $this->connection->createInstance([
+            'parent' => $projectName,
+            'instanceId' => $this->id,
+            'instance' => [
+                'displayName' => $displayName,
+                'type' => $type,
+                'labels' => $labels
+            ],
+            'clusters' => $clustersArray
+        ] + $options);
+
+        return $this->resumeOperation($operation['name'], $operation);
+    }
+
+    /**
+     * This method returns an associative array containing data required for
+     * configuring Bigtable instance clusters. The resulting array will contain
+     * 'clusterId', 'locationId`, and depending on how it is called,
+     * `storageType` and `serveNodes`.
+     *
+     * Example:
+     * ```
+     * $cluster = Instance::buildClusterMetadata('my-cluster', 'us-east1-b');
+     * ```
+     *
+     * @codingStandardsIgnoreStart
+     * @param string $clusterId The cluster ID
+     *        e.g., just `cluster-id` rather than `projects/project-id/instances/instance-id/clusters/cluster-id`.
+     * @param string $locationId The location ID
+     *        e.g., just `us-east1-b` rather than `projects/project-id/locations/us-east1-b`.
+     * @param int $storageType The storage media type for persisting Bigtable data. Possible values include
+     *        `Google\Cloud\Bigtable\Admin\V2\StorageType::SSD`,
```

Comment:
The constants living on this class were great, why did we remove them?

Reply:
didnt see much value of creating constants out of another constant, better to use same constant everywhere

Reply:
We [originally asked for this](https://github.com/GoogleCloudPlatform/google-cloud-php/pull/1086#discussion_r197812527) because it means users don't need to have knowledge of the generated classes in order to successfully use the veneer.

Reply:
ok, but I think we are providing enough documentation for user to have this information handy. After merging all gapic code as part of this same veneer client I dont think it really matters whether the class referenced is generated or handwritten. [Instance_Type](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/Bigtable/src/Admin/V2/Instance_Type.php) and [StorageType](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/Bigtable/src/Admin/V2/StorageType.php)

Reply:
We are trying to stick to standards that have already been set. Consistency is important to us, and we want users who are used to using our other veneers to feel at home. That is why we are suggesting the same path we took for Spanner.

Reply:
I dont think this is exactly same scenario as spanner Instance, The constants defined in Spanner instance is only used/referred from Spanner Instance and output only. While constants used in this case are Inputs and will need to be referred from multiple files like BigtableClient when we add createInstance in BigtableClient. In this case its better to refer to from actual common class rather than referring to wrapper.

Reply:
> For eg StorageType is property of Cluster, having this Constant in Instance doesnt make sense.

We would eventually have a `Cluster` class here as well, why wouldn't the constant just live there?

Reply:
Bcz this Cluster constant will be used/referred from three different class after that, Cluster, Instance and Bigtable. It definitely makes better standard to use values from the class dedicated to Constants rather than using it from a wrapper created else where. 

One other problem with this standard is that when there are multiple Constants properties for eg Instance will have State and Type, Cluster will have State & StorageType, this will make it confusing for user and error prone too.

Reply:
Our strategy throughout all of the veneer clients over the past three years has always been to make proto classes and values implementation details. If a user is not directly using the generated client, they should not need any knowledge at all of protobuf.

The number of constants in a class is totally immaterial. Each constant is documented in context on the methods where it is applicable, so whether there is one constant or 20, it makes little difference.

I understand your concerns, but this is an area where we have a very established convention which has served us well for a long time. Our job as developers of a library is to provide, as much as possible, consistent clients which allow users of one client to easily work with another client, taking their experience and expectations with them, and implementing APIs in a manner that conforms with existing expectations.

We know that this review has been difficult, and we're extremely close to finishing it. I don't want this to be a point of ongoing contention that holds up our ability to move on to other areas of the client, but unfortunately David and I are very firm in our requirements for this.

Reply:
@jdpedrie, this constants change is going to affect a lot of files.  Can we please merge this as as, and then block any other merges until we fix the constants issue?

Reply:
That will be fine.

Reply:
@dwsupplee: thanks.  Refreshing this page is a bit painful right now with all of the comments :)